File: scaffold/src/main/java/top/xuqingquan/web/publics/AgentWebUtils.java
package top.xuqingquan.web.publics;

import android.app.Activity;
import android.content.Context;
import android.os.Build;
import android.os.Handler;
import android.text.SpannableString;
import android.text.Spanned;
import android.text.TextUtils;
import android.text.style.ForegroundColorSpan;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewParent;
import androidx.annotation.ColorInt;
import androidx.annotation.RequiresApi;
import com.google.android.material.snackbar.Snackbar;
import top.xuqingquan.R;
import top.xuqingquan.utils.FileUtils;
import top.xuqingquan.utils.Timber;
import top.xuqingquan.web.nokernel.PermissionInterceptor;
import top.xuqingquan.web.nokernel.WebConfig;
import top.xuqingquan.web.nokernel.WebUtils;

import java.io.File;
import java.lang.ref.WeakReference;

import static top.xuqingquan.web.nokernel.WebUtils.isUIThread;

public class AgentWebUtils {

    private AgentWebUtils() {
        throw new UnsupportedOperationException("u can't init me");
    }

    static void clearWebView(android.webkit.WebView m) {
        if (m == null) {
            return;
        }
        if (!isUIThread()) {
            return;
        }
        m.loadUrl("about:blank");
        m.stopLoading();
        if (m.getHandler() != null) {
            m.getHandler().removeCallbacksAndMessages(null);
        }
        m.removeAllViews();
        ViewGroup mViewGroup = ((ViewGroup) m.getParent());
        if (mViewGroup != null) {
            mViewGroup.removeView(m);
        }
        m.setWebChromeClient(null);
        m.setWebViewClient(null);
        m.setTag(null);
        m.clearHistory();
        m.destroy();
    }

    static void clearWebView(com.tencent.smtt.sdk.WebView m) {
        if (m == null) {
            return;
        }
        if (!isUIThread()) {
            return;
        }
        m.loadUrl("about:blank");
        m.stopLoading();
        if (m.getHandler() != null) {
            m.getHandler().removeCallbacksAndMessages(null);
        }
        m.removeAllViews();
        ViewGroup mViewGroup = ((ViewGroup) m.getParent());
        if (mViewGroup != null) {
            mViewGroup.removeView(m);
        }
        m.setWebChromeClient(null);
        m.setWebViewClient(null);
        m.setTag(null);
        m.clearHistory();
        m.destroy();
    }

    public static void show(View parent,
                            CharSequence text,
                            int duration,
                            @ColorInt int textColor,
                            @ColorInt int bgColor,
                            CharSequence actionText,
                            @ColorInt int actionTextColor,
                            View.OnClickListener listener) {
        SpannableString spannableString = new SpannableString(text);
        ForegroundColorSpan colorSpan = new ForegroundColorSpan(textColor);
        spannableString.setSpan(colorSpan, 0, spannableString.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
        WeakReference<Snackbar> snackbarWeakReference = new WeakReference<>(Snackbar.make(parent, spannableString, duration));
        Snackbar snackbar = snackbarWeakReference.get();
        View view = snackbar.getView();
        view.setBackgroundColor(bgColor);
        if (actionText != null && actionText.length() > 0 && listener != null) {
            snackbar.setActionTextColor(actionTextColor);
            snackbar.setAction(actionText, listener);
        }
        snackbar.show();
    }

    public static void clearWebViewAllCache(Context context, android.webkit.WebView webView) {
        try {
            AgentWebConfig.removeAllCookies(null);
            webView.getSettings().setCacheMode(android.webkit.WebSettings.LOAD_NO_CACHE);
            context.deleteDatabase("webviewCache.db");
            context.deleteDatabase("webview.db");
            webView.clearCache(true);
            webView.clearHistory();
            webView.clearFormData();
            FileUtils.clearCacheFolder(new File(WebUtils.getCachePath(context)), 0);
        } catch (Throwable t) {
            Timber.e(t);
        }
    }

    public static void clearWebViewAllCache(Context context, com.tencent.smtt.sdk.WebView webView) {
        try {
            AgentWebConfig.removeAllX5Cookies(null);
            webView.getSettings().setCacheMode(android.webkit.WebSettings.LOAD_NO_CACHE);
            context.deleteDatabase("webviewCache.db");
            context.deleteDatabase("webview.db");
            webView.clearCache(true);
            webView.clearHistory();
            webView.clearFormData();
            FileUtils.clearCacheFolder(new File(WebUtils.getCachePath(context)), 0);
        } catch (Throwable t) {
            Timber.e(t);
        }
    }

    public static void clearWebViewAllCache(Context context) {
        try {
            if (WebConfig.hasX5()) {
                clearWebViewAllCache(context, new com.tencent.smtt.sdk.WebView(context.getApplicationContext()));
            } else {
                clearWebViewAllCache(context, new android.webkit.WebView(context.getApplicationContext()));
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static AbsAgentWebUIController getAgentWebUIControllerByWebView(android.webkit.WebView webView) {
        WebParentLayout mWebParentLayout = getWebParentLayoutByWebView(webView);
        return mWebParentLayout.provide();
    }

    public static AbsAgentWebUIController getAgentWebUIControllerByWebView(com.tencent.smtt.sdk.WebView webView) {
        WebParentLayout mWebParentLayout = getWebParentLayoutByWebView(webView);
        return mWebParentLayout.provide();
    }

    static WebParentLayout getWebParentLayoutByWebView(android.webkit.WebView webView) {
        ViewGroup mViewGroup;
        if (!(webView.getParent() instanceof ViewGroup)) {
            throw new IllegalStateException("please check webcreator's create method was be called ?");
        }
        mViewGroup = (ViewGroup) webView.getParent();
        while (mViewGroup != null) {

            Timber.i("ViewGroup:" + mViewGroup);
            if (mViewGroup.getId() == R.id.web_parent_layout_id) {
                WebParentLayout mWebParentLayout = (WebParentLayout) mViewGroup;
                Timber.i("found WebParentLayout");
                return mWebParentLayout;
            } else {
                ViewParent mViewParent = mViewGroup.getParent();
                if (mViewParent instanceof ViewGroup) {
                    mViewGroup = (ViewGroup) mViewParent;
                } else {
                    mViewGroup = null;
                }
            }
        }
        throw new IllegalStateException("please check webcreator's create method was be called ?");
    }

    static WebParentLayout getWebParentLayoutByWebView(com.tencent.smtt.sdk.WebView webView) {
        ViewGroup mViewGroup;
        if (!(webView.getParent() instanceof ViewGroup)) {
            throw new IllegalStateException("please check webcreator's create method was be called ?");
        }
        mViewGroup = (ViewGroup) webView.getParent();
        while (mViewGroup != null) {

            Timber.i("ViewGroup:" + mViewGroup);
            if (mViewGroup.getId() == R.id.web_parent_layout_id) {
                WebParentLayout mWebParentLayout = (WebParentLayout) mViewGroup;
                Timber.i("found WebParentLayout");
                return mWebParentLayout;
            } else {
                ViewParent mViewParent = mViewGroup.getParent();
                if (mViewParent instanceof ViewGroup) {
                    mViewGroup = (ViewGroup) mViewParent;
                } else {
                    mViewGroup = null;
                }
            }
        }
        throw new IllegalStateException("please check webcreator's create method was be called ?");
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    public static boolean showFileChooserCompat(Activity activity,
                                                android.webkit.WebView webView,
                                                android.webkit.ValueCallback valueCallbacks,
                                                android.webkit.WebChromeClient.FileChooserParams fileChooserParams,
                                                PermissionInterceptor permissionInterceptor,
                                                android.webkit.ValueCallback valueCallback,
                                                String mimeType,
                                                Handler.Callback jsChannelCallback
    ) {
        try {
            FileChooser.Builder builder = FileChooser.newBuilder(activity, webView);
            if (valueCallbacks != null) {
                builder.setUriValueCallbacks(valueCallbacks);
            }
            if (fileChooserParams != null) {
                builder.setFileChooserParams(fileChooserParams);
            }
            if (valueCallback != null) {
                builder.setUriValueCallback(valueCallback);
            }
            if (!TextUtils.isEmpty(mimeType)) {
                builder.setAcceptType(mimeType);
            }
            if (jsChannelCallback != null) {
                builder.setJsChannelCallback(jsChannelCallback);
            }
            builder.setPermissionInterceptor(permissionInterceptor);
            builder.build().openFileChooser();
        } catch (Throwable throwable) {
            Timber.e(throwable);
            if (valueCallbacks != null) {
                Timber.i("onReceiveValue empty");
                return false;
            }
            if (valueCallback != null) {
                valueCallback.onReceiveValue(null);
            }
        }
        return true;
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    public static boolean showFileChooserCompat(Activity activity,
                                                com.tencent.smtt.sdk.WebView webView,
                                                com.tencent.smtt.sdk.ValueCallback valueCallbacks,
                                                com.tencent.smtt.sdk.WebChromeClient.FileChooserParams fileChooserParams,
                                                PermissionInterceptor permissionInterceptor,
                                                com.tencent.smtt.sdk.ValueCallback valueCallback,
                                                String mimeType,
                                                Handler.Callback jsChannelCallback
    ) {
        try {
            FileChooser.Builder builder = FileChooser.newBuilder(activity, webView);
            if (valueCallbacks != null) {
                builder.setUriValueCallbacks(valueCallbacks);
            }
            if (fileChooserParams != null) {
                builder.setFileChooserParams(fileChooserParams);
            }
            if (valueCallback != null) {
                builder.setUriValueCallback(valueCallback);
            }
            if (!TextUtils.isEmpty(mimeType)) {
                builder.setAcceptType(mimeType);
            }
            if (jsChannelCallback != null) {
                builder.setJsChannelCallback(jsChannelCallback);
            }
            builder.setPermissionInterceptor(permissionInterceptor);
            builder.build().openFileChooser();
        } catch (Throwable throwable) {
            Timber.e(throwable);
            if (valueCallbacks != null) {
                Timber.i("onReceiveValue empty");
                return false;
            }
            if (valueCallback != null) {
                valueCallback.onReceiveValue(null);
            }
        }
        return true;
    }
}

File: scaffold/src/main/java/top/xuqingquan/web/publics/AgentWebUtils.kt
package top.xuqingquan.web.publics

import android.app.Activity
import android.content.Context
import android.net.Uri
import android.os.Build
import android.os.Handler
import android.text.SpannableString
import android.text.Spanned
import android.text.TextUtils
import android.text.style.ForegroundColorSpan
import android.view.View
import android.view.ViewGroup
import androidx.annotation.ColorInt
import androidx.annotation.RequiresApi
import com.google.android.material.snackbar.Snackbar
import top.xuqingquan.R
import top.xuqingquan.utils.Timber
import top.xuqingquan.utils.clearCacheFolder
import top.xuqingquan.web.nokernel.PermissionInterceptor
import top.xuqingquan.web.nokernel.WebConfig
import top.xuqingquan.web.nokernel.WebUtils
import top.xuqingquan.web.nokernel.WebUtils.isUIThread
import java.io.File
import java.lang.ref.WeakReference

object AgentWebUtils {

    @JvmStatic
    fun clearWebView(m: android.webkit.WebView?) {
        try {
            if (m == null) {
                return
            }
            if (!isUIThread()) {
                return
            }
            m.loadUrl("about:blank")
            m.stopLoading()
            if (m.handler != null) {
                m.handler.removeCallbacksAndMessages(null)
            }
            m.removeAllViews()
            val mViewGroup = m.parent as ViewGroup?
            mViewGroup?.removeView(m)
            m.webChromeClient = null
            m.webViewClient = null
            m.tag = null
            m.clearHistory()
            m.destroy()
        } catch (t: Throwable) {
        }
    }

    @JvmStatic
    fun clearWebView(m: com.tencent.smtt.sdk.WebView?) {
        try {
            if (m == null) {
                return
            }
            if (!isUIThread()) {
                return
            }
            m.loadUrl("about:blank")
            m.stopLoading()
            if (m.handler != null) {
                m.handler.removeCallbacksAndMessages(null)
            }
            m.removeAllViews()
            val mViewGroup = m.parent as ViewGroup?
            mViewGroup?.removeView(m)
            m.webChromeClient = null
            m.webViewClient = null
            m.tag = null
            m.clearHistory()
            m.destroy()
        } catch (t: Throwable) {
        }
    }

    @JvmStatic
    fun show(
        parent: View,
        text: CharSequence,
        duration: Int,
        @ColorInt textColor: Int,
        @ColorInt bgColor: Int,
        actionText: CharSequence?,
        @ColorInt actionTextColor: Int,
        listener: View.OnClickListener?
    ) {
        val spannableString = SpannableString(text)
        val colorSpan = ForegroundColorSpan(textColor)
        spannableString.setSpan(
            colorSpan,
            0,
            spannableString.length,
            Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
        )
        val snackbarWeakReference = WeakReference(Snackbar.make(parent, spannableString, duration))
        val snackbar = snackbarWeakReference.get() ?: return
        snackbar.view.setBackgroundColor(bgColor)
        if (!actionText.isNullOrEmpty() && listener != null) {
            snackbar.setActionTextColor(actionTextColor)
            snackbar.setAction(actionText, listener)
        }
        snackbar.show()
    }

    @JvmStatic
    fun clearWebViewAllCache(context: Context, webView: android.webkit.WebView) {
        try {
            AgentWebConfig.removeAllCookies(null)
            webView.settings.cacheMode = android.webkit.WebSettings.LOAD_NO_CACHE
            context.deleteDatabase("webviewCache.db")
            context.deleteDatabase("webview.db")
            webView.clearCache(true)
            webView.clearHistory()
            webView.clearFormData()
            clearCacheFolder(File(WebUtils.getCachePath(context)), 0)
        } catch (t: Throwable) {
            Timber.e(t)
        }

    }

    @JvmStatic
    fun clearWebViewAllCache(context: Context, webView: com.tencent.smtt.sdk.WebView) {
        try {
            AgentWebConfig.removeAllX5Cookies(null)
            webView.settings.cacheMode = android.webkit.WebSettings.LOAD_NO_CACHE
            context.deleteDatabase("webviewCache.db")
            context.deleteDatabase("webview.db")
            webView.clearCache(true)
            webView.clearHistory()
            webView.clearFormData()
            clearCacheFolder(File(WebUtils.getCachePath(context)), 0)
        } catch (t: Throwable) {
            Timber.e(t)
        }

    }

    @JvmStatic
    fun clearWebViewAllCache(context: Context) {
        try {
            if (WebConfig.hasX5()) {
                clearWebViewAllCache(
                    context,
                    top.xuqingquan.web.x5.LollipopFixedWebView(context.applicationContext)
                )
            } else {
                clearWebViewAllCache(context, top.xuqingquan.web.system.LollipopFixedWebView(context.applicationContext))
            }
        } catch (e: Throwable) {
            e.printStackTrace()
        }

    }

    @JvmStatic
    fun getAgentWebUIControllerByWebView(webView: android.webkit.WebView): AbsAgentWebUIController? {
        val mWebParentLayout = getWebParentLayoutByWebView(webView)
        return mWebParentLayout.provide()
    }

    @JvmStatic
    fun getAgentWebUIControllerByWebView(webView: com.tencent.smtt.sdk.WebView): AbsAgentWebUIController? {
        val mWebParentLayout = getWebParentLayoutByWebView(webView)
        return mWebParentLayout.provide()
    }

    @JvmStatic
    fun getWebParentLayoutByWebView(webView: android.webkit.WebView): WebParentLayout {
        var mViewGroup: ViewGroup?
        check(webView.parent is ViewGroup) { "please check webcreator's create method was be called ?" }
        mViewGroup = webView.parent as ViewGroup
        while (mViewGroup != null) {
            Timber.i("ViewGroup:$mViewGroup")
            mViewGroup = if (mViewGroup.id == R.id.scaffold_web_parent_layout_id) {
                val mWebParentLayout = mViewGroup as WebParentLayout?
                Timber.i("found WebParentLayout")
                return mWebParentLayout!!
            } else {
                val mViewParent = mViewGroup.parent
                if (mViewParent is ViewGroup) {
                    mViewParent
                } else {
                    null
                }
            }
        }
        throw IllegalStateException("please check webcreator's create method was be called ?")
    }

    @JvmStatic
    fun getWebParentLayoutByWebView(webView: com.tencent.smtt.sdk.WebView): WebParentLayout {
        var mViewGroup: ViewGroup?
        check(webView.parent is ViewGroup) { "please check webcreator's create method was be called ?" }
        mViewGroup = webView.parent as ViewGroup
        while (mViewGroup != null) {
            Timber.i("ViewGroup:$mViewGroup")
            mViewGroup = if (mViewGroup.id == R.id.scaffold_web_parent_layout_id) {
                val mWebParentLayout = mViewGroup as WebParentLayout?
                Timber.i("found WebParentLayout")
                return mWebParentLayout!!
            } else {
                val mViewParent = mViewGroup.parent
                if (mViewParent is ViewGroup) {
                    mViewParent
                } else {
                    null
                }
            }
        }
        throw IllegalStateException("please check webcreator's create method was be called ?")
    }

    @JvmStatic
    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    fun showFileChooserCompat(
        activity: Activity,
        webView: android.webkit.WebView,
        valueCallbacks: android.webkit.ValueCallback<Array<Uri>>?,
        fileChooserParams: android.webkit.WebChromeClient.FileChooserParams?,
        permissionInterceptor: PermissionInterceptor,
        valueCallback: android.webkit.ValueCallback<Uri>?,
        mimeType: String?,
        jsChannelCallback: Handler.Callback?
    ): Boolean {
        try {
            val builder = FileChooser.newBuilder(activity, webView)
            if (valueCallbacks != null) {
                builder.setUriValueCallbacks(valueCallbacks)
            }
            if (fileChooserParams != null) {
                builder.setFileChooserParams(fileChooserParams)
            }
            if (valueCallback != null) {
                builder.setUriValueCallback(valueCallback)
            }
            if (!TextUtils.isEmpty(mimeType)) {
                builder.setAcceptType(mimeType)
            }
            if (jsChannelCallback != null) {
                builder.setJsChannelCallback(jsChannelCallback)
            }
            builder.setPermissionInterceptor(permissionInterceptor)
            builder.build().openFileChooser()
        } catch (throwable: Throwable) {
            Timber.e(throwable)
            if (valueCallbacks != null) {
                Timber.i("onReceiveValue empty")
                return false
            }
            valueCallback?.onReceiveValue(null)
        }

        return true
    }

    @JvmStatic
    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    fun showFileChooserCompat(
        activity: Activity,
        webView: com.tencent.smtt.sdk.WebView,
        valueCallbacks: com.tencent.smtt.sdk.ValueCallback<Array<Uri>>?,
        fileChooserParams: com.tencent.smtt.sdk.WebChromeClient.FileChooserParams?,
        permissionInterceptor: PermissionInterceptor,
        valueCallback: com.tencent.smtt.sdk.ValueCallback<Uri>?,
        mimeType: String?,
        jsChannelCallback: Handler.Callback?
    ): Boolean {
        try {
            val builder = FileChooser.newBuilder(activity, webView)
            if (valueCallbacks != null) {
                builder.setUriValueCallbacks(valueCallbacks)
            }
            if (fileChooserParams != null) {
                builder.setFileChooserParams(fileChooserParams)
            }
            if (valueCallback != null) {
                builder.setUriValueCallback(valueCallback)
            }
            if (!TextUtils.isEmpty(mimeType)) {
                builder.setAcceptType(mimeType)
            }
            if (jsChannelCallback != null) {
                builder.setJsChannelCallback(jsChannelCallback)
            }
            builder.setPermissionInterceptor(permissionInterceptor)
            builder.build().openFileChooser()
        } catch (throwable: Throwable) {
            Timber.e(throwable)
            if (valueCallbacks != null) {
                Timber.i("onReceiveValue empty")
                return false
            }
            valueCallback?.onReceiveValue(null)
        }

        return true
    }
}
